helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,774 @@
+package storagepacker
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	radix "github.com/armon/go-radix"
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/errwrap"
+	log "github.com/hashicorp/go-hclog"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/sdk/helper/compressutil"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"github.com/hashicorp/vault/sdk/helper/locksutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/sdk/physical"
+)
+
+const (
+	defaultBaseBucketBits  = 8
+	defaultBucketShardBits = 4
+)
+
+var (
+	shardLocks = make(map[*locksutil.LockEntry]struct{}, 32)
+)
+
+type Config struct {
+	// BucketStorageView is the storage to be used by all the buckets
+	BucketStorageView *logical.StorageView `json:"-"`
+
+	// ConfigStorageView is the storage to store config info
+	ConfigStorageView *logical.StorageView `json:"-"`
+
+	// Logger for output
+	Logger log.Logger `json:"-"`
+
+	// BaseBucketBits is the number of bits to use for buckets at the base level
+	BaseBucketBits int `json:"base_bucket_bits"`
+
+	// BucketShardBits is the number of bits to use for sub-buckets a bucket
+	// gets sharded into when it reaches the maximum threshold.
+	BucketShardBits int `json:"-"`
+}
+
+// StoragePacker packs many items into abstractions called buckets. The goal
+// is to employ a reduced number of storage entries for a relatively huge
+// number of items. This is the second version of the utility which supports
+// indefinitely expanding the capacity of the storage by sharding the buckets
+// when they exceed the imposed limit.
+type StoragePackerV2 struct {
+	*Config
+	storageLocks []*locksutil.LockEntry
+	bucketsCache *radix.Tree
+
+	// Note that we're slightly loosy-goosy with this lock. The reason is that
```

Comment:
I'd generally prefer a more concrete description of how the lock should be used. I checked through the review and I didn't find cases where a method was called on bucketsCache that *didn't* have the lock, so I'm confused.